### PR TITLE
[EOC-424] Database model migrations: list cohorts and users not working

### PR DESCRIPTION
**Pull request by alemikolo**
This branch started from EOC_421.
To implement DB migration I use [mongo-migrate](https://github.com/seppevs/migrate-mongo#readme) library.
Migration files are saved in `server/migrations` directory.
Migration state is saved in mongo in `migrations` collection.

To run the migrations enter: `npm run migrate up`

### CODE REVIEW

File: migrate-mongo-config.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const config = {
+  mongodb: {
+    url: 'mongodb://localhost:27017',
```
Comment (marekrozmus) [inline migrate-mongo-config.js]: take this from 'common/variables'

File: server/migrations/20190821104046-0.0.1.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    )
+    .then(() =>
+      db
+        .collection('lists')
+        .updateMany(
+          { items: { $exists: true, $ne: [] } },
+          { $set: { 'items.$[].isDeleted': false } }
+        )
+    )
+    .catch(() => {
```
Comment (marekrozmus) [inline server/migrations/20190821104046-0.0.1.js]: remove braces

File: server/migrations/20190821104046-0.0.1.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    )
+    .then(() =>
+      db
+        .collection('lists')
+        .updateMany(
+          { items: { $exists: true, $ne: [] } },
+          { $unset: { 'items.$[].isDeleted': false } }
+        )
+    )
+    .catch(() => {
```
Comment (marekrozmus) [inline server/migrations/20190821104046-0.0.1.js]: remove braces

File: server/migrations/20190821122149-0.0.2.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    )
+    .then(() =>
+      db
+        .collection('lists')
+        .updateMany(
+          { items: { $exists: true, $ne: [] } },
+          { $set: { 'items.$[].locks': locks } }
+        )
+    )
+    .catch(() => {
```
Comment (marekrozmus) [inline server/migrations/20190821122149-0.0.2.js]: remove braces

File: server/migrations/20190821122149-0.0.2.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    )
+    .then(() =>
+      db
+        .collection('lists')
+        .updateMany(
+          { items: { $exists: true, $ne: [] } },
+          { $unset: { 'items.$[].locks': {} } }
+        )
+    )
+    .catch(() => {
```
Comment (marekrozmus) [inline server/migrations/20190821122149-0.0.2.js]: remove braces

File: server/migrations/20190821125125-0.0.3.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+const Settings = require('../models/settings.model');
+
+const up = db =>
+  db
+    .collection('users')
+    .updateMany(
+      { settings: { $exists: false } },
+      { $set: { settings: new Settings() } }
+    )
+    .catch(() => {
```
Comment (marekrozmus) [inline server/migrations/20190821125125-0.0.3.js]: remove braces

File: server/migrations/20190821125125-0.0.3.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+      console.log('up migration failed...');
+    });
+
+const down = db =>
+  db
+    .collection('users')
+    .updateMany({ settings: { $exists: true } }, { $unset: { settings: {} } })
+    .catch(() => {
```
Comment (marekrozmus) [inline server/migrations/20190821125125-0.0.3.js]: remove braces

File: migrate-mongo-config.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    databaseName: 'eoc',
```
Comment (adamklepacz) [inline migrate-mongo-config.js]: this can be also taken from common/variables

File: server/migrations/20190821104046-0.0.1.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
+/* eslint-disable no-console */
+const up = db =>
+  db
```
Comment (Ciunkos) [inline server/migrations/20190821104046-0.0.1.js]:
```javascript
const task1 = () => db
        .collection('lists')
        .updateMany(
          { isDeleted: { $exists: false } },
          { $set: { isDeleted: false } }
        )

...

const tasks = [task1, task2, task3];

tasks.forEach(task => task())

Promise.all(tasks).then("done").catch("error");
```

File: package.json [hunk 10]
```diff
@@ -14,6 +14,7 @@
     "lint": "eslint ./ --ext \".js,.jsx\" --max-warnings=0",
     "lint:fix": "eslint ./ --ext \".js,.jsx\" --max-warnings=0 --fix",
     "migrate": "migrate-mongo",
+    "migrate:up": "migrate-mongo up",
```
Comment (adamklepacz) [inline package.json]: maybe add also script for migrating down

File: server/common/utils/helpers.js [hunk 11]
```diff
@@ -326,6 +326,17 @@ const responseWithCohortDetails = (doc, userId) => {
...
 };
 
+const runAsyncTasks = async (...tasks) => {
+  tasks.forEach(task => task());
+
+  try {
+    await Promise.all(tasks);
+  } catch {
+    // eslint-disable-next-line no-console
+    console.log('migration failed...');
```
Comment (adamklepacz) [inline server/common/utils/helpers.js]: Migration failed...

start with capital letter :)

File: server/tests/setupTest.js [hunk 12]
```diff
@@ -0,0 +1 @@
+require('babel-polyfill');
```
Comment (adamklepacz) [inline server/tests/setupTest.js]: what is this??

Comment (alemikolo) [reply server/tests/setupTest.js]: Helper which I added use async-await. And `Jest` doesn't recognize those statements even if I didn't write any test for it (maybe it analyzes all files which it has to test). So I found some solutions: https://stackoverflow.com/questions/42535270/regeneratorruntime-is-not-defined-when-running-jest-test and with this line `jest` doesn't throw any errors...

File: server/common/utils/helpers.js [hunk 13]
```diff
@@ -326,6 +326,17 @@ const responseWithCohortDetails = (doc, userId) => {
...
 };
 
+const runAsyncTasks = async (...tasks) => {
+  tasks.forEach(task => task());
+
+  try {
+    await Promise.all(tasks);
+  } catch {
+    // eslint-disable-next-line no-console
+    console.log('Migration failed...');
```
Comment (Ciunkos) [inline server/common/utils/helpers.js]: Please `console.error` and rethrow.

File: server/migrations/20190821104046-0.0.1.js [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+  const removeIsDeletedFromItems = async () =>
+    db
+      .collection('lists')
+      .updateMany(
+        { items: { $exists: true, $ne: [] } },
+        { $unset: { 'items.$[].isDeleted': false } }
+      );
+
+  runAsyncTasks(
```
Comment (Ciunkos) [inline server/migrations/20190821104046-0.0.1.js]: Please await this to catch errors.

File: server/migrations/20190821122149-0.0.2.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+  const addLocksToItems = async () =>
+    db
+      .collection('lists')
+      .updateMany(
+        { items: { $exists: true, $ne: [] } },
+        { $set: { 'items.$[].locks': locks } }
+      );
+
+  runAsyncTasks(addLocksToCohorts, addLocksToCLists, addLocksToItems);
```
Comment (Ciunkos) [inline server/migrations/20190821122149-0.0.2.js]: Please await all `runAsyncTasks` calls.

File: README.md [hunk 16]
```diff
@@ -96,6 +98,11 @@ To run MongoDB locally you need to clone all the files from this repository and
 By default MongoDB server will run at `localhost:27017`.
 There might be others mongodb proccesses runnig at this port. To stop all runing mongodb services type in your terminal: <br/> `sudo service mongodb stop`. Than simply rerun `sudo mongod`.
 
+
+### DB migrations
+
+To perform DB migrations we use [migrate-mongo](https://github.com/seppevs/migrate-mongo#readme) tool. To config migrate-mongo edit `migrate-mongo-config.js` file. To create migration file run `npm run migrate create <version>` in your terminal. Migration file will be created at `server/migrations` directory. To run all new migrations run `npm run migrate:up` or `npm run migrate up` in your terminal. To rollback last migration run `npm run migrate:down` or `npm run migrate down` in your terminal. To see migrations status run `npm run migrate status` in your terminal.
```
Comment (Ciunkos) [inline README.md]: Shouldn't it be `npm run migrate -- up` and `npm run migrate -- down`?

File: server/common/utils/helpers.js [hunk 17]
```diff
@@ -326,6 +326,18 @@ const responseWithCohortDetails = (doc, userId) => {
...
 };
 
+const runAsyncTasks = async (...tasks) => {
+  tasks.forEach(task => task());
+
+  try {
+    await Promise.all(tasks);
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.error('Migration failed...');
```
Comment (Ciunkos) [inline server/common/utils/helpers.js]: It would be nice to log the error as well to ease the debugging with `console.error(error)`.

**Pull request closed without merging**